### DISCUSSION TITLE
V7 - what is a good alternative to previously blocking triggerValidation method?

### DISCUSSION THREAD

**Discussion by fantapop** [react-hook-form/react-hook-form#4448] [Q&A]
For various reasons, I'm unable to rely on the form submission directly so I'm using a button to submit the data to our backend.  In an old version of react-hook-form we had some code like this: 

```typescript
    const onSubmitPress = useCallback(() => {
        const isValid = await triggerValidation();
        if (isValid) {
            // continue with backend submission
        }
    });
```

Is there a new recommended way to do this now that `trigger` is non-blocking?

**Comment by bluebill1049** [comment]
i think you should use `handleSubmit` instead of `trigger`.

**Reply by deviprsd** [reply]
I would say it is still not a viable solution, though `handleSubmit` provides a blocking way it takes away the functionality of `trigger` on specific fields. Another use case of blocking trigger would be **wizard forms**. IMO `trigger` should return a `Promise<void>`

**Reply by bluebill1049** [reply]
i mean you can subscribe to the errors object as an alternative (by calling trigger).

**Reply by Motoxpro** [reply]
Looking for a solution here as well. I was executing individual mutations so I could save fields as the user completed the form (HUGE form so no way to only save at the end, only other solution would be a ton of useForms(40+))

```    
const handleCurrencyCodeSelect = async (): Promise<void> => {
      const isValidCurrencyCode = await trigger('currencyCode');
      if (isValidCurrencyCode) {
        updateCurrencyCodeMutation(getValues('currencyCode'));
      }
    };
```

Can I do this now?

```
const handleCurrencyCodeSelect = (): void => {
      trigger('currencyCode');
      if (errors.currencyCode) {
        return;
      }
      updateCurrencyCodeMutation(getValues('currencyCode'));
    };
```

**Reply by Motoxpro** [reply]
Wrote a function that solves this

```
/**
 * RHF V7 doesn't allow us to get a return on if the field is valid or not
 * We can use this function to trigger the validation and observe the error
 */
export async function blockingIsValidTrigger<T>(
  formMethods: UseFormReturn<T>,
  name: string,
): Promise<boolean> {
  const {
    formState: { errors },
    trigger,
  } = formMethods;
  await trigger(name);
  return !errors?.[name];
}
```

**Reply by deviprsd** [reply]
@Motoxpro How are you awaiting on `trigger` when trigger isn't a `Promise`?

**Reply by bluebill1049** [reply]
We are bring it back in the next release.

**Reply by Motoxpro** [reply]
@deviprsd If await is used on a non-promised function, that function is wrapped in a promise. See https://stackoverflow.com/a/55263084/7435898

In this case, we don't need the return value, we just need to ensure the execution order and that the function has fired before we get the value of `errors`.

**Reply by deviprsd** [reply]
@Motoxpro I would emphasize on this 
> Therefore, await changes the execution order (but you should not rely on it nevertheless)

My obvious concern is that even if it wrapped in a promise it isn't necessarily resolved at the point of completion, lets test that theory

```javascript
function z() {
    setTimeout(() => console.log(123), 5000);
}

async function x() {
    await z();
    console.log(456);
}
```

Upon running `x()` the output was

> 456
> 123

Now your code perhaps works because trigger probably doesn't take that long and await changes the execution order a little, had it taken a little longer `1s`, `2s` ... this approach would not work.

Obviously, that said, a promise version for it **has been planned for the next release!**

**Reply by bluebill1049** [reply]
The previous trigger is brought back.

**Reply by Motoxpro** [reply]
Thank you for the education @deviprsd and thank you @bluebill1049 for bringing it back! :)
